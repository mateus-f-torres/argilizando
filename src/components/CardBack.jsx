//@flow
import * as React from 'react';
import {Link} from 'react-router-dom';

type Props = {
  name: string,
  pitch: string,
  skill: Array<string>,
  tool?: Array<string>,
  lang?: Array<number | string>,
  equip: Array<Array<string>>,
  lockBack: () => void
};

const CardBack = (props: Props) => (
  <figure className="card">
    <h2>{props.name}</h2>
    <p>{props.pitch}</p>
    <div id="hide" className="horizontal-line" />
    <h3>Background Features</h3>
    <ul id="skills" className="multi">
      <h4>Skill Proficiency</h4>
      {props.skill.map((sk, i) => <li key={i}>{sk}</li>)}
    </ul>

    {
      props.tool &&
        <ul id="tools" className="multi">
          <h4>Tool Proficiency</h4>
          {props.tool.map((item, i) => <li key={i}>{item}</li>)}
        </ul>
    }

    {
      props.lang &&
        <ul id="langs" className="multi">
          <h4>Languages</h4>
          <li>{props.lang[1]}</li>
        </ul>
    }

    <ul id="equip" className="dotted-list">
      <h4>Equipment</h4>
      {
        props.equip.map((list, i) => (
          <li key={i}>
            {list.map((item, j) => <span key={j}>{item}<br/></span>)}
          </li>
        ))
      }
    </ul>
    <div className="horizontal-line no-bottom" />
    <div className="lock-box">
      <Link to="/background">
        <button>Back</button>
      </Link>
      <Link to="/">
        <button onClick={props.lockBack}>Lock</button>
      </Link>
    </div>
  </figure>
);

export default CardBack;
