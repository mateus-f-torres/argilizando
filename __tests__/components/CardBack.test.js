import React from 'react';
import {shallow} from 'enzyme';
import CardBack from '../../src/components/CardBack.jsx';
import BACKGROUNDS from '../../src/data/backgrounds.js';

describe('<CardBack />', () => {
  describe('Unit Tests', () => {
    describe('Background with Tools and Languages', () => {
      let back = {
        name: 'Noble',
        pitch: `You understand wealth, power, and privilege.`,
        skill: ['History', 'Persuasion'],
        tool: ['One gaming set'],
        lang: [1, 'One of your choice'],
        equip: [
          ['set of fine clothes'],
          ['signet ring'],
          ['scroll of pedigree'],
          ['purse containing 25gp'],
        ],
      };

      it('displays correct background name', () => {
        const wrapper = shallow(<CardBack {...back}/>);

        expect(wrapper.find('[data-test="back-name"]')
          .text()).toMatch(/Noble/);
      });

      it('displays correct background pitch', () => {
        const wrapper = shallow(<CardBack {...back}/>);

        expect(wrapper.find('[data-test="back-pitch"]')
          .text()).toMatch(new RegExp(back.pitch));
      });

      it('displays correct background skills list', () => {
        const wrapper = shallow(<CardBack {...back}/>);

        expect(wrapper.find('[data-test="back-skill"]')
          // 2 skills (li tags) + 1 header (h4 tag)
          .children().length).toBe(3);
      });

      it('displays correct background tools list', () => {
        const wrapper = shallow(<CardBack {...back}/>);

        expect(wrapper.find('[data-test="back-tool"]')
          // 1 tool (li tag) + 1 header (h4 tag)
          .children().length).toBe(2);
      });

      it('displays correct background extra language text', () => {
        const wrapper = shallow(<CardBack {...back}/>);

        expect(wrapper.find('[data-test="back-lang"]')
          // how many langs text (li tag) + 1 header (h4 tag)
          .children().length).toBe(2);

        expect(wrapper.find('[data-test="back-lang"]')
          .childAt(1).text()).toMatch(/One/);
      });

      it('displays correct equipment list', () => {
        const wrapper = shallow(<CardBack {...back}/>);
        expect(wrapper.find('[data-test="back-equip"]')
          // 3 traits (li tags) + 1 header (h3 tag)
          .children().length).toEqual(5);
      });
    });

    describe('Background without Tools', () => {
      let back = {
        name: 'Acolyte',
        pitch: `You understand wealth, power, and privilege.`,
        skill: ['History', 'Persuasion'],
        lang: [2, 'Two of your choice'],
        equip: [
          ['holy symbol'],
          ['prayer book', 'or prayer wheel'],
          ['5 sticks of incense'],
          ['vestments'],
          ['set of common clothes'],
          ['belt pouch containing 15gp'],
        ],
      };

      it('displays correct background name', () => {
        const wrapper = shallow(<CardBack {...back}/>);

        expect(wrapper.find('[data-test="back-name"]')
          .text()).toMatch(/Acolyte/);
      });

      it('tools list not shown', () => {
        const wrapper = shallow(<CardBack {...back}/>);

        expect(wrapper.find('[data-test="back-tool"]')
          .exists()).toBe(false);
      });

      it('displays correct background extra language text', () => {
        const wrapper = shallow(<CardBack {...back}/>);

        expect(wrapper.find('[data-test="back-lang"]')
          // how many langs text (li tag) + 1 header (h4 tag)
          .children().length).toBe(2);

        expect(wrapper.find('[data-test="back-lang"]')
          .childAt(1).text()).toMatch(/Two/);
      });
    });

    describe('Background without Languages', () => {
      let back = {
        name: 'Charlatan',
        pitch: `You have always had a way with people.`,
        skill: ['Deception', 'Sleight of Hand'],
        tool: ['Disguise kit', 'Forgery kit'],
        equip: [
          ['set of fine clothes'],
          ['diguise kit'],
          ['10 bottles with colored liquid',
            'or set of weighted dice',
            'or deck of marked cards',
            'or ring of an imaginary duke'],
          ['belt pouch containing 15g'],
        ],
      };

      it('displays correct background name', () => {
        const wrapper = shallow(<CardBack {...back}/>);

        expect(wrapper.find('[data-test="back-name"]')
          .text()).toMatch(/Charlatan/);
      });

      it('language text not shown', () => {
        const wrapper = shallow(<CardBack {...back}/>);

        expect(wrapper.find('[data-test="back-lang"]')
          .exists()).toBe(false);
      });

      it('displays correct background tools list', () => {
        const wrapper = shallow(<CardBack {...back}/>);

        expect(wrapper.find('[data-test="back-tool"]')
          // 2 tools (li tags) + 1 header (h4 tag)
          .children().length).toBe(3);
      });
    });
  });

  describe('Snapshot Test', () => {
    it('displays correct noble background card', () => {
      // 1st category -> options -> noble
      let back = BACKGROUNDS[0][1][2];

      const wrapper = shallow(<CardBack {...back}/>);

      expect(wrapper).toMatchSnapshot();
    });
  });
});
