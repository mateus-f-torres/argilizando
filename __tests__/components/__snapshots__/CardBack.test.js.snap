// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardBack /> Snapshot Test displays correct noble background card 1`] = `
<figure
  className="card"
>
  <h2
    data-test="back-name"
  >
    Noble
  </h2>
  <p
    data-test="back-pitch"
  >
    You understand wealth, power, and privilege. 
        You carry a noble title, and your family owns land,
        collects taxes, and wields significant political influence.
  </p>
  <div
    className="horizontal-line"
    id="hide"
  />
  <h3>
    Background Features
  </h3>
  <ul
    className="multi"
    data-test="back-skill"
    id="skills"
  >
    <h4>
      Skill Proficiency
    </h4>
    <li
      key="0"
    >
      History
    </li>
    <li
      key="1"
    >
      Persuasion
    </li>
  </ul>
  <ul
    className="multi"
    data-test="back-tool"
    id="tools"
  >
    <h4>
      Tool Proficiency
    </h4>
    <li
      key="0"
    >
      One gaming set
    </li>
  </ul>
  <ul
    className="multi"
    data-test="back-lang"
    id="langs"
  >
    <h4>
      Languages
    </h4>
    <li>
      One of your choice
    </li>
  </ul>
  <ul
    className="dotted-list"
    data-test="back-equip"
    id="equip"
  >
    <h4>
      Equipment
    </h4>
    <li
      key="0"
    >
      <span
        key="0"
      >
        set of fine clothes
        <br />
      </span>
    </li>
    <li
      key="1"
    >
      <span
        key="0"
      >
        signet ring
        <br />
      </span>
    </li>
    <li
      key="2"
    >
      <span
        key="0"
      >
        scroll of pedigree
        <br />
      </span>
    </li>
    <li
      key="3"
    >
      <span
        key="0"
      >
        purse containing 25gp
        <br />
      </span>
    </li>
  </ul>
  <div
    className="horizontal-line no-bottom"
  />
  <div
    className="lock-box"
  >
    <Link
      replace={false}
      to="/background"
    >
      <button>
        Back
      </button>
    </Link>
    <Link
      replace={false}
      to="/"
    >
      <button>
        Lock
      </button>
    </Link>
  </div>
</figure>
`;
